src/index/CompressedRelation.cpp

```diff
@@ -513,15 +513,76 @@ size_t CompressedRelationReader::getResultSizeOfScan(
   return numResults;
 }
 
-// _____________________________________________________________________________
+// ____________________________________________________________________________
+IdTable CompressedRelationReader::getDistinctCol1IdsAndCounts(
+    const CompressedRelationMetadata& relationMetadata,
+    const std::vector<CompressedBlockMetadata>& allBlocksMetadata) const {
+  IdTableStatic<2> table(allocator_);
+  std::span<const CompressedBlockMetadata> relationBlocksMetadata =
+      getBlocksFromMetadata(relationMetadata, std::nullopt, allBlocksMetadata);
+
+  // Iterate over the blocks and only read (and decompress) those which
+  // contain more than one different `col1Id`. For the others, we can determine
+  // the count from the metadata.
+  std::optional<Id> currentCol1Id;
+  size_t currentCount = 0;
+  size_t rowIndex = std::numeric_limits<size_t>::max();
+  // Helper lambda that processes a single `col1Id` (with `howMany`
+  // occurrences). If it's new, a row with the previous `col1Id` and
+  // `currentCount` is added to `table`, and `currentCol1Id` and `currentCount`
+  // are reset.
+  auto processCol1Id = [&table, &rowIndex, &currentCol1Id, &currentCount](
+                           std::optional<Id> col1Id, size_t howMany) {
+    if (col1Id != currentCol1Id) {
+      if (currentCol1Id.has_value()) {
+        table.emplace_back();
+        table(rowIndex, 0) = currentCol1Id.value();
+        table(rowIndex, 1) = Id::makeFromInt(currentCount);
+        rowIndex++;
```

Comment:
```suggestion
        table.push_back({currentCol1Id.value(), Id::makeFromInt(...)});
```
(There one of the overloads of `push_back` should make this a one-liner that gets rid of the `rowIndex` completely.